pallets/randomness/src/lib.rs

```diff
@@ -0,0 +1,434 @@
+						log::warn!("Could not read BABE two epochs ago randomness from the relay");
+					}
+				}
+			}
+			<RelayTime<T>>::put(RelayTimeInfo {
+				relay_block_number,
+				relay_epoch_index,
+			});
+			<InherentIncluded<T>>::put(());
+
+			Ok(Pays::No.into())
+		}
+	}
+
+	#[pallet::inherent]
+	impl<T: Config> ProvideInherent for Pallet<T> {
+		type Call = Call<T>;
+		type Error = InherentError;
+		const INHERENT_IDENTIFIER: InherentIdentifier = INHERENT_IDENTIFIER;
+
+		fn is_inherent_required(_: &InherentData) -> Result<Option<Self::Error>, Self::Error> {
+			// Return Ok(Some(_)) unconditionally because this inherent is required in every block
+			// If it is not found, throw a VrfInherentRequired error.
+			Ok(Some(InherentError::Other(
+				sp_runtime::RuntimeString::Borrowed(
+					"Inherent required to set babe randomness results",
+				),
+			)))
+		}
+
+		// The empty-payload inherent extrinsic.
+		fn create_inherent(_data: &InherentData) -> Option<Self::Call> {
+			Some(Call::set_babe_randomness_results {})
+		}
+
+		fn is_inherent(call: &Self::Call) -> bool {
+			matches!(call, Call::set_babe_randomness_results { .. })
+		}
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		// Set this block's randomness using the VRF output, verified by the VrfInput put in
+		// storage in the previous block's `on_initialize`
+		fn on_initialize(_now: BlockNumberFor<T>) -> Weight {
+			// Set and validate VRF output using input put in storage last block's `on_finalize`
+			vrf::set_output::<T>()
+		}
+		// Set next block's VRF input in storage
+		fn on_finalize(_now: BlockNumberFor<T>) {
+			// Panics if set_babe_randomness_results inherent was not included
+			assert!(
+				<InherentIncluded<T>>::take().is_some(),
+				"Mandatory randomness inherent not included; InherentIncluded storage item is empty"
+			);
+
+			// Necessary because required data is killed in `ParachainSystem::on_initialize`
+			// which may happen before the VRF output is verified in the next block on_initialize
+			vrf::set_input::<T>();
+		}
+	}
+
+	// Utility function
+	impl<T: Config> Pallet<T> {
+		pub(crate) fn concat_and_hash(a: T::Hash, b: H256) -> [u8; 32] {
+			let mut s = Vec::new();
+			s.extend_from_slice(a.as_ref());
+			s.extend_from_slice(b.as_ref());
+			sp_io::hashing::blake2_256(&s)
+		}
+	}
+
+	// Public functions for precompile usage only
+	impl<T: Config> Pallet<T> {
+		pub fn request_randomness(request: Request<T>) -> DispatchResult {
+			ensure!(
+				!request.can_be_fulfilled(),
+				Error::<T>::CannotRequestPastRandomness
+			);
+			let total_to_reserve = T::Deposit::get().saturating_add(request.fee);
+			let contract_address =
+				T::AddressMapping::into_account_id(request.contract_address.clone());
+			T::ReserveCurrency::can_reserve(&contract_address, total_to_reserve)
+				.then(|| true)
+				.ok_or(Error::<T>::InsufficientDeposit)?;
+			// get new request ID
+			let request_id = <RequestCount<T>>::get();
```

Comment:
This might cause some weird bugs since the first `request_id` is `0` which is semantically correct (so are negative numbers as ids for that matter) but is not quite intuitive to the average use-case of testing for existence. Might wanna start the id at `1` by using the `next_id` instead.

```typescrtipt
if (request_id) {
    // happy path
}
```

Reply:
@girazoki @librelois do you have an opinion here?

Reply:
I think request_id 0 is fine, I am of the opinion that this should be sufficiently intuitive for SC developers, they should be used to this already.

Reply:
Recall that this will be used by SCs more likely than end users.